Avoid using the default system Locale and printing to System.out in production code

leventov opened the pull request:
- Avoid using the default system locale, use `Locale.ENGLISH` instead:
   - In `String.format()`, replace with `StringUtils.format()` in the whole codebase: 90% changes of this PR.
   - `String.toLowerCase()` and `toUpperCase()`, replace with `StringUtils.toLowerCase()` and `toUpperCase()` accordingly.
   - `System.out.printf()`, in test code
   - Misc

This is not just stylistic change, it may actually lead to errors if Druid-running servers have non-English locale, [`String.toLowerCase()` Javadoc gives a scary example](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#toLowerCase--).

 - Don't use `System.out`, `e.printStackTrace()` and similar in production code, replace with logging. Except for a few classes in `druid-services`, like `io.druid.cli.Main`.
 - Replace `new IOException(String.format(...))` with `new IOE(...)`, and similarly with `ISE`, `IAE` and `UOE` where possible.
 - A small fix: https://github.com/druid-io/druid/pull/4326#issuecomment-306568658

This PR doesn't enforce prohibition of using the default system locale and `System.out`, because I want to do that with this plugin: https://github.com/policeman-tools/forbidden-apis, and it also forbids using the default timezone, for what I want to create a separate PR.

Fixes #4335

This PR is a copy of #4392 which was accidentially closed by branch removal.

leventov commented:
Could anyone please review this? It blocks timezone unification PR that might help with #4126

drcrallen commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByMergingQueryRunnerV2.java:
```diff
@@ -139,7 +140,7 @@ public GroupByMergingQueryRunnerV2(
     final File temporaryStorageDirectory = new File(
         processingTmpDir,
-        String.format("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
+        StringUtils.safeFormat("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
```
Probably should crash if bad format string.

drcrallen commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -79,7 +80,7 @@
     final File temporaryStorageDirectory = new File(
         processingTmpDir,
-        String.format("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
+        StringUtils.safeFormat("druid-groupBy-%s_%s", UUID.randomUUID(), query.getId())
```
probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedTemporaryStorage.java:
```diff
@@ -81,7 +82,7 @@ public LimitedOutputStream createFile() throws IOException
       FileUtils.forceMkdir(storageDirectory);
 
-      final File theFile = new File(storageDirectory, String.format("%08d.tmp", files.size()));
+      final File theFile = new File(storageDirectory, StringUtils.safeFormat("%08d.tmp", files.size()));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/query/spec/SpecificSegmentQueryRunner.java:
```diff
@@ -63,7 +64,7 @@ public SpecificSegmentQueryRunner(
     final Thread currThread = Thread.currentThread();
     final String currThreadName = currThread.getName();
-    final String newName = String.format("%s_%s_%s", query.getType(), query.getDataSource(), query.getIntervals());
+    final String newName = StringUtils.safeFormat("%s_%s_%s", query.getType(), query.getDataSource(), query.getIntervals());
```
should probably crash on bad format

drcrallen commented on processing/src/main/java/io/druid/segment/FloatColumnSerializer.java:
```diff
@@ -64,7 +65,7 @@ public void open() throws IOException
     writer = CompressionFactory.getFloatSerializer(
         ioPeon,
-        String.format("%s.float_column", filenameBase),
+        StringUtils.safeFormat("%s.float_column", filenameBase),
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/FloatMetricColumnSerializer.java:
```diff
@@ -57,7 +58,7 @@ public FloatMetricColumnSerializer(
   {
     writer = CompressionFactory.getFloatSerializer(
-        ioPeon, String.format("%s_little", metricName), IndexIO.BYTE_ORDER, compression
+        ioPeon, StringUtils.safeFormat("%s_little", metricName), IndexIO.BYTE_ORDER, compression
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -544,7 +546,7 @@ public void convertV8toV9(File v8Dir, File v9Dir, IndexSpec indexSpec)
             ByteBuffer dimBuffer = v8SmooshedFiles.mapFile(filename);
             String dimension = serializerUtils.readString(dimBuffer);
-            if (!filename.equals(String.format("dim_%s.drd", dimension))) {
+            if (!filename.equals(StringUtils.safeFormat("dim_%s.drd", dimension))) {
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -693,7 +695,7 @@ public int size()
             // legacy merger (which will be deprecated) to support long/float dims. Going forward, the V9 merger
             // should be used instead if any dimension types beyond String are needed.
-            if (!filename.endsWith(String.format("%s.drd", BYTE_ORDER))) {
+            if (!filename.endsWith(StringUtils.safeFormat("%s.drd", BYTE_ORDER))) {
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -741,7 +743,7 @@ public int size()
             final ColumnDescriptor serdeficator = builder.build();
             makeColumn(v9Smoosher, metric, serdeficator);
-          } else if (String.format("time_%s.drd", BYTE_ORDER).equals(filename)) {
+          } else if (StringUtils.safeFormat("time_%s.drd", BYTE_ORDER).equals(filename)) {
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -1047,21 +1049,21 @@ private Column deserializeColumn(ObjectMapper mapper, ByteBuffer byteBuffer, Smo
   public static File makeDimFile(File dir, String dimension)
   {
-    return new File(dir, String.format("dim_%s.drd", dimension));
+    return new File(dir, StringUtils.safeFormat("dim_%s.drd", dimension));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -1047,21 +1049,21 @@ private Column deserializeColumn(ObjectMapper mapper, ByteBuffer byteBuffer, Smo
   public static File makeNumericDimFile(File dir, String dimension, ByteOrder order)
   {
-    return new File(dir, String.format("numeric_dim_%s_%s.drd", dimension, order));
+    return new File(dir, StringUtils.safeFormat("numeric_dim_%s_%s.drd", dimension, order));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -1047,21 +1049,21 @@ private Column deserializeColumn(ObjectMapper mapper, ByteBuffer byteBuffer, Smo
   public static File makeTimeFile(File dir, ByteOrder order)
   {
-    return new File(dir, String.format("time_%s.drd", order));
+    return new File(dir, StringUtils.safeFormat("time_%s.drd", order));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexIO.java:
```diff
@@ -1047,21 +1049,21 @@ private Column deserializeColumn(ObjectMapper mapper, ByteBuffer byteBuffer, Smo
   public static File makeMetricFile(File dir, String metricName, ByteOrder order)
   {
-    return new File(dir, String.format("met_%s_%s.drd", metricName, order));
+    return new File(dir, StringUtils.safeFormat("met_%s_%s.drd", metricName, order));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -648,7 +650,7 @@ public Metadata apply(IndexableAdapter input)
 
         dataInterval = new Interval(minTime, maxTime);
-        serializerUtils.writeString(channel, String.format("%s/%s", minTime, maxTime));
+        serializerUtils.writeString(channel, StringUtils.safeFormat("%s/%s", minTime, maxTime));
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -825,7 +827,7 @@ public String apply(@Nullable String input)
             formatString = "dim_%s.drd";
           } else {
-            formatString = String.format("numeric_dim_%%s_%s.drd", IndexIO.BYTE_ORDER);
+            formatString = StringUtils.safeFormat("numeric_dim_%%s_%s.drd", IndexIO.BYTE_ORDER);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -834,11 +836,11 @@ public String apply(@Nullable String input)
           Iterables.concat(
               Arrays.asList(
-                  "index.drd", "inverted.drd", "spatial.drd", String.format("time_%s.drd", IndexIO.BYTE_ORDER)
+                  "index.drd", "inverted.drd", "spatial.drd", StringUtils.safeFormat("time_%s.drd", IndexIO.BYTE_ORDER)
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -834,11 +836,11 @@ public String apply(@Nullable String input)
               Iterables.transform(mergedDimensions, dimFilenameFunction),
               Iterables.transform(
-                  mergedMetrics, GuavaUtils.formatFunction(String.format("met_%%s_%s.drd", IndexIO.BYTE_ORDER))
+                  mergedMetrics, GuavaUtils.formatFunction(StringUtils.safeFormat("met_%%s_%s.drd", IndexIO.BYTE_ORDER))
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -1044,7 +1046,7 @@ public void createIndexDrdFile(
       availableMetrics.writeToChannel(channel);
       serializerUtils.writeString(
-          channel, String.format("%s/%s", dataInterval.getStart(), dataInterval.getEnd())
+          channel, StringUtils.safeFormat("%s/%s", dataInterval.getStart(), dataInterval.getEnd())
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/LongColumnSerializer.java:
```diff
@@ -71,7 +72,7 @@ public void open() throws IOException
     writer = CompressionFactory.getLongSerializer(
         ioPeon,
-        String.format("%s.long_column", filenameBase),
+        StringUtils.safeFormat("%s.long_column", filenameBase),
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/LongMetricColumnSerializer.java:
```diff
@@ -61,7 +62,7 @@ public LongMetricColumnSerializer(
   {
     writer = CompressionFactory.getLongSerializer(
-        ioPeon, String.format("%s_little", metricName), IndexIO.BYTE_ORDER, encoding, compression
+        ioPeon, StringUtils.safeFormat("%s_little", metricName), IndexIO.BYTE_ORDER, encoding, compression
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerLegacy.java:
```diff
@@ -126,7 +127,7 @@ public void close() throws IOException
     boolean hasSpatial = capabilities.hasSpatialIndexes();
     if (hasSpatial) {
-      String spatialFilename = String.format("%s.spatial", dimensionName);
+      String spatialFilename = StringUtils.safeFormat("%s.spatial", dimensionName);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerLegacy.java:
```diff
@@ -95,7 +96,7 @@ public void writeIndexes(List<IntBuffer> segmentRowNumConversions, Closer closer
     final BitmapSerdeFactory bitmapSerdeFactory = indexSpec.getBitmapSerdeFactory();
 
-    String bmpFilename = String.format("%s.inverted", dimensionName);
+    String bmpFilename = StringUtils.safeFormat("%s.inverted", dimensionName);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -156,7 +157,7 @@ public void writeMergedValueMetadata(List<IndexableAdapter> adapters) throws IOE
     }
 
-    String dictFilename = String.format("%s.dim_values", dimensionName);
+    String dictFilename = StringUtils.safeFormat("%s.dim_values", dimensionName);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -200,7 +201,7 @@ protected void setupEncodedValueWriter() throws IOException
     final CompressedObjectStrategy.CompressionStrategy compressionStrategy = indexSpec.getDimensionCompression();
 
-    String filenameBase = String.format("%s.forward_dim", dimensionName);
+    String filenameBase = StringUtils.safeFormat("%s.forward_dim", dimensionName);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -275,7 +276,7 @@ public void writeIndexes(List<IntBuffer> segmentRowNumConversions, Closer closer
     final BitmapSerdeFactory bitmapSerdeFactory = indexSpec.getBitmapSerdeFactory();
 
-    String bmpFilename = String.format("%s.inverted", dimensionName);
+    String bmpFilename = StringUtils.safeFormat("%s.inverted", dimensionName);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java:
```diff
@@ -309,7 +310,7 @@ public void close()
         spatialWriter = new ByteBufferWriter<>(
             ioPeon,
-            String.format("%s.spatial", dimensionName),
+            StringUtils.safeFormat("%s.spatial", dimensionName),
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/ByteBufferWriter.java:
```diff
@@ -76,7 +77,7 @@ public void write(T objectToWrite) throws IOException
   private String makeFilename(String suffix)
   {
-    return String.format("%s.%s", filenameBase, suffix);
+    return StringUtils.safeFormat("%s.%s", filenameBase, suffix);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/CompressedVSizeIndexedV3Writer.java:
```diff
@@ -48,14 +49,14 @@ public static CompressedVSizeIndexedV3Writer create(
         new CompressedIntsIndexedWriter(
             ioPeon,
-            String.format("%s.offsets", filenameBase),
+            StringUtils.safeFormat("%s.offsets", filenameBase),
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/CompressedVSizeIndexedV3Writer.java:
```diff
@@ -48,14 +49,14 @@ public static CompressedVSizeIndexedV3Writer create(
         new CompressedVSizeIntsIndexedWriter(
             ioPeon,
-            String.format("%s.values", filenameBase),
+            StringUtils.safeFormat("%s.values", filenameBase),
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/EntireLayoutIndexedLongSupplier.java:
```diff
@@ -59,7 +60,7 @@ public void fill(int index, long[] toFill)
       if (totalSize - index < toFill.length) {
         throw new IndexOutOfBoundsException(
-            String.format(
+            StringUtils.safeFormat(
```
do you want to just add an `IOBE`?

leventov replied:
This pattern is not used too many times and it's not as immediately understandable abbr for Java devs as NPE or ISE, so no

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -236,7 +236,7 @@ private void checkIndex(int index)
     }
     if (index >= size) {
-      throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+      throw new IAE(StringUtils.safeFormat("Index[%s] >= size[%s]", index, size));
```
don't need to wrap safeFormat

leventov replied:
Fixed

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -63,8 +63,8 @@
  * <p>
- * Header file name is identified as:  String.format("%s_header", columnName)
- * value files are identified as: String.format("%s_value_%d", columnName, fileNumber)
+ * Header file name is identified as: StringUtils.safeFormat("%s_header", columnName)
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -63,8 +63,8 @@
- * Header file name is identified as:  String.format("%s_header", columnName)
- * value files are identified as: String.format("%s_value_%d", columnName, fileNumber)
+ * Header file name is identified as: StringUtils.safeFormat("%s_header", columnName)
+ * value files are identified as: StringUtils.safeFormat("%s_value_%d", columnName, fileNumber)
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java:
```diff
@@ -97,12 +97,12 @@ public GenericIndexedWriter(
   public static String generateValueFileName(String fileNameBase, int fileNum)
   {
-    return String.format("%s_value_%d", fileNameBase, fileNum);
+    return StringUtils.safeFormat("%s_value_%d", fileNameBase, fileNum);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java:
```diff
@@ -97,12 +97,12 @@ public GenericIndexedWriter(
   public static String generateHeaderFileName(String fileNameBase)
   {
-    return String.format("%s_header", fileNameBase);
+    return StringUtils.safeFormat("%s_header", fileNameBase);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java:
```diff
@@ -160,7 +160,7 @@ public void write(T objectToWrite) throws IOException
   private String makeFilename(String suffix)
   {
-    return String.format("%s.%s", filenameBase, suffix);
+    return StringUtils.safeFormat("%s.%s", filenameBase, suffix);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/VSizeIndexedIntsWriter.java:
```diff
@@ -51,7 +52,7 @@ public VSizeIndexedIntsWriter(
   {
     this.ioPeon = ioPeon;
-    this.valueFileName = String.format("%s.values", filenameBase);
+    this.valueFileName = StringUtils.safeFormat("%s.values", filenameBase);
```
Probably should crash if bad format string

drcrallen commented on processing/src/main/java/io/druid/segment/data/VSizeIndexedWriter.java:
```diff
@@ -64,9 +65,9 @@ public VSizeIndexedWriter(
-    this.metaFileName = String.format("%s.meta", filenameBase);
-    this.headerFileName = String.format("%s.header", filenameBase);
-    this.valuesFileName = String.format("%s.values", filenameBase);
+    this.metaFileName = StringUtils.safeFormat("%s.meta", filenameBase);
```
Probably should crash if bad format string

leventov commented:
@drcrallen thanks for this catch. I think not only for file names, but in many and probably sometimes hardly unpredictable cases it may be important not to emit something wrong instead of failing, so I renamed `safeFormat()` to `unimportantSafeFormat()` and use it only for logging and exception messages, and added `StringUtils.format()` which simply delegates to `String.format()` with `Locale.ENGLISH`, which now used everywhere else.

leventov commented:
BTW it this `safeFormat()` was protection from programming errors, IntelliJ has inspections which verify format strings and arguments of any format-like methods, that could be used to effectively exclude such errors. The few places where it doesn't apply are `StringFormatExtractionFn` and `IndexerSQLMetadataStorageCoordinator`.

leventov commented:
@drcrallen what do you think about the proposed solution?

drcrallen commented on java-util/src/main/java/io/druid/java/util/common/StringUtils.java:
```diff
@@ -100,7 +100,20 @@ public static String fromUtf8(final ByteBuffer buffer)
+   * concatenated format string and format arguments. Should be used for unimportant formatting like logging,
+   * exception messages, typically not directly.
+   */
+  public static String unimportantSafeFormat(String message, Object... formatArgs)
```
suggest simply naming `logFormat` to indicate it should only be used for log-level string creation for stuff that is not structured.

leventov replied:
It is also used in Exception constructors, like IOE

drcrallen replied:
yes but those are just logging style messages (unstructured)

drcrallen replied:
basically "unimportant" is not correct. "unstructured" is probably more correct, or something similar

gianm replied:
My 2¢: the "safe" was originally meant to convey the behavior, so if it fails at this, then it doesn't need to stay. Consider something like `nonStrictFormat` or `lenientFormat` (since "non-strict" and "lenient" probably convey the behavior better than "safe").

leventov replied:
Renamed to nonStrictFormat because unstructured IMO is confusing

leventov commented:
@drcrallen or @gianm could this PR be merged now?
